lib/libzfs_core/libzfs_core.c

```diff
@@ -171,6 +179,41 @@ libzfs_core_fini(void)
 	(void) pthread_mutex_unlock(&g_lock);
 }
 
+int
+lzc_ioctl_fd(int fd, unsigned long ioc, zfs_cmd_t *zc)
+{
+	if (!g_ioc_trace)
+		return (lzc_ioctl_fd_os(fd, ioc, zc));
+
+	nvlist_t *nvl;
+
+	fprintf(stderr, "=== lzc_ioctl: call: ioc=0x%lx name=%s\n",
+	    ioc, zc->zc_name[0] ? zc->zc_name : "[none]");
+	if (zc->zc_nvlist_src) {
+		nvl = fnvlist_unpack(
+		    (void *)(uintptr_t)zc->zc_nvlist_src,
+		    zc->zc_nvlist_src_size);
+		nvlist_print(stderr, nvl);
+		fnvlist_free(nvl);
+	}
+
+	int rc = lzc_ioctl_fd_os(fd, ioc, zc);
+	int err = errno;
+
+	fprintf(stderr, "=== lzc_ioctl: result: ioc=0x%lx name=%s errno=%d\n",
+	    ioc, zc->zc_name[0] ? zc->zc_name : "[none]", errno);
```

Comment:
I would check `rc` here too to reduce possible noise in logs.
```suggestion
	    ioc, zc->zc_name[0] ? zc->zc_name : "[none]", (rc < 0) ? err : 0);
```

Reply:
I wouldn't be against including `rc=%d` in the output too since it indicates how `errno` should be interpreted.  However, I don't insist on it as long as you pull in the suggested change here.

Reply:
I've gone with both, it's a good idea. Fragment from `zfs list`:

```
...
=== lzc_ioctl: call: ioc=0x5a14 name=crayon/home/robn
=== lzc_ioctl: result: ioc=0x5a14 name=crayon/home/robn/ rc=-1 errno=3
=== lzc_ioctl: call: ioc=0x5a14 name=crayon/home
=== lzc_ioctl: result: ioc=0x5a14 name=crayon/home/root rc=-1 errno=12
=== lzc_ioctl: call: ioc=0x5a14 name=crayon/home
=== lzc_ioctl: result: ioc=0x5a14 name=crayon/home/root rc=0 errno=0
nvlist version: 0
	mountpoint = (embedded nvlist)
	nvlist version: 0
		value = /root
		source = crayon/home/root
	(end mountpoint)
...
```